src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -140,22 +169,61 @@ class LazyCachingPersistenceStore[K, Category, Serialized](
     val category = ir.category
     val storageId = ir.toStorageId(id, None)
     lockManager.executeSequentially(category.toString) {
-      async { // linter:ignore UnnecessaryElseBranch
-        await(store.store(id, v, version))
-        val old = idCache.getOrElse(category, Nil) // linter:ignore UndesirableTypeInference
-        idCache.put(category, id +: old)
-        Done
+      lockManager.executeSequentially(storageId.toString) {
+        async { // linter:ignore UnnecessaryElseBranch
+          await(store.store(id, v, version))
+          updateCachedVersions(storageId, version, category, v)
+          Done
+        }
       }
     }
   }
 
-  override def versions[Id, V](id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] =
-    store.versions(id)
+  protected[cache] def maybePurgeCachedVersions(toRemove: Int = 500, pRemoval: Double = 0.05)(f: () => Boolean): Unit =
+    while (f()) {
+      // randomly GC the versions
+      val counter = new AtomicInteger
+      versionedValueCache.retain { (k, v) =>
+        val x = scala.util.Random.nextDouble()
```

Comment:
as discussed via slack: detecting "at max capacity" and simply removing 1 element at random is not very efficient because all subsequent writes will also be at max capacity. better to remove (at random) some percentage of existing entries.